File: Google Codejam/2013/Round 1B/Falling Diamonds/main.py
#!/usr/bin/python
def completed(N):
    A, add, level = 1, 1, 0
    while A < N:
        add += 4
        if A + add > N:
            return level, N - A
        A += add
        level += 1
    return level, 0


MAXN = 3000
comb = [[0] * MAXN for i in range(MAXN)]
comb[0][0] = 1
for i in range(MAXN - 1):
    for j in range(MAXN - 1):
        comb[i + 1][j] += comb[i][j]
        comb[i + 1][j + 1] += comb[i][j]

from decimal import *
T = int(input())
for test in range(1, T + 1):
    N, X, Y = [int(x) for x in input().split()]

    LEVELS, LEFT = completed(N)
    LEVELS *= 2

    ME = abs(X) + abs(Y)
    answer = None
    if ME <= LEVELS:
        answer = Decimal(1)
    elif ME > LEVELS + 2:
        answer = Decimal(0)
    elif X == 0 or LEFT == 0:
        answer = Decimal(0)
    else:
        need = Y + 1

        total = 0
        good = 0
        for i in range(LEFT + 1):
            j = LEFT - i
            if i > LEVELS + 2 or j > LEVELS + 2:
                continue
            total += comb[LEFT][i]
            if i >= need:
                good += comb[LEFT][i]

        answer = Decimal(good) / Decimal(total)

    print("Case #" + str(test) + ":", answer)

File: Google Codejam/2013/Round 1B/Falling Diamonds/main.cpp
#include <iostream>
#include <fstream>
#include <algorithm>

using namespace std;

pair<int, int> completed(int N) {
    int A = 1, add = 1, level = 0;
    while (A < N) {
        add += 4;
        if (A + add > N)
            return {level, N - A};
        A += add;
        level += 1;
    }
    return {level, 0};
}

int abs(int x) {
    if (x < 0)
        return -x;
    return x;
}

int main() {
    int T; cin >> T;
    for (int test = 1; test <= T; ++test) {
        int N, X, Y; cin >> N >> X >> Y;

        int LEVELS, LEFT;
        tie(LEVELS, LEFT) = completed(N);
        LEVELS *= 2;
        double answer = 0;
        int ME = abs(X) + abs(Y);

        if (ME <= LEVELS)
            answer = 1;
        else if (ME > LEVELS + 2)
            answer = 0;
        else if (X == 0 or LEFT == 0)
            answer = 0;
        else {
            int need = Y + 1;

            LEVELS += 2;
            vector< vector<double> > dp(LEVELS + 1, vector<double>(LEVELS + 1, 0));
            dp[0][0] = 1;
            answer = 0;
            for (int i = 0; i <= LEVELS; ++i)
                for (int j = 0; j <= LEVELS; ++j) {
                    if (i + j == LEFT and i >= need)
                        answer += dp[i][j];
 
                    if (i == LEVELS and j == LEVELS)
                        continue;
                    if (i == LEVELS)
                        dp[i][j + 1] += dp[i][j];
                    else if (j == LEVELS)
                        dp[i + 1][j] += dp[i][j];
                    else {
                        dp[i + 1][j] += dp[i][j] / 2;
                        dp[i][j + 1] += dp[i][j] / 2;
                    }

               }
        }

        cout.setf(ios::fixed, ios::floatfield);
        cout.precision(10);
        cout << "Case #" << test << ": " << answer << "\n";
    }
}
